### ISSUE TITLE
Docker cp not overwriting directories on volumes

### ISSUE THREAD

**Issue by lazysoundsystem** [docker/docker#25755]
Seems to be a change from 1.10 to 1.12 - a script that worked fine before is now not copying some files, apparently to directories that have been created on the container, but I need to dig deeper...

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:11:10 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:11:10 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 32
 Running: 8
 Paused: 0
 Stopped: 24
Images: 44
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 82
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay host null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.697 GiB
Name: andy-Notebook
ID: EE62:24XC:RERS:OEK2:UDZ4:V6RQ:XSFS:C4V3:L6RZ:QVCD:MPIB:F42Y
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Local environment.

**Steps to reproduce the issue:**

```
andy@andy-Notebook:~/Desktop$ find one -type f
one/two/three/three
one/two/three/four/four
one/two/two
one/one
```

```
andy@andy-Notebook:~/Desktop$ docker create --name test -it alpine sh
25ce54fba950b66bfd316aae52cd6e1d5f30ea590789780aaf5eea50dda96021
andy@andy-Notebook:~/Desktop$ docker cp one test:/
andy@andy-Notebook:~/Desktop$ docker start -ai test
/ # find /one -type f
/one/one
/one/two/three/three
/one/two/three/four/four
/one/two/two
```

```
andy@andy-Notebook:~/Desktop$ docker rm test
test
andy@andy-Notebook:~/Desktop$ docker create --name test -it -v /one alpine sh
fb9c8b6a77b50635db464f5dec706a7a17fd8b507949a212985e8d5ac4cc2b15
andy@andy-Notebook:~/Desktop$ docker cp one test:/
andy@andy-Notebook:~/Desktop$ docker start -ai test
/ # find /one -type f
/ #
```

```
andy@andy-Notebook:~/Desktop$ docker rm test
test
andy@andy-Notebook:~/Desktop$ docker create --name test -it -v /one/two alpine sh
d7ae36b79794190e2adb3954b1c3c51283e781176bed765cf1c6be9eeab3122a
andy@andy-Notebook:~/Desktop$ docker cp one test:/
andy@andy-Notebook:~/Desktop$ docker start -ai test
/ # find /one -type f
/one/one
/ #
```

**Describe the results you received:**
This example looks like it's just not overwriting directories (in Docker 1.10, I get the same output for each example). 
I've had other issues - it sometimes copies only directories but not files, I'll try to add examples later.

**Describe the results you expected:**
I expected all the files in the `one` directory to be present in the container.

**Additional information you deem important (e.g. issue happens only occasionally):**
Not consistent behavior - I'll get back to you on this!

**Comment by justincormack** [comment]
I don't think `docker cp` has ever copied files onto or off volumes as in your second two examples, it copies to or from the root filesystem. That can be confusing. It is not clear from your example though - you do not show the difference with 1.10.

**Comment by lazysoundsystem** [comment]
Thanks for looking at this - I'm also not sure there are any guarantees of that behavior in the docs https://docs.docker.com/engine/reference/commandline/cp/

This is how it works for me in Docker 1.10:
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Sat Mar 12 19:18:57 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Sat Mar 12 19:18:57 2016
 OS/Arch:      linux/amd64
```

```
[lazysoundsystem@sd-44415 ~]$ docker create --name test -it alpine sh
830d92bc089410420197480cdb926338abc56b0de7a0ca68b07dea45b9aa1c4f
[lazysoundsystem@sd-44415 ~]$ docker cp one test:/
[lazysoundsystem@sd-44415 ~]$ docker start -ai test
/ # find /one -type f
/one/one
/one/two/three/four/four
/one/two/three/three
/one/two/two
/ #
```

```
/ # [lazysoundsystem@sd-44415 ~]$ docker rm test
test
[lazysoundsystem@sd-44415 ~]$ docker create --name test -it -v /one alpine sh
6c0edfe30043b644af88951ad76e6d64496751b53a65500a5ed01bcb2e414d1d
[lazysoundsystem@sd-44415 ~]$ docker cp one test:/
[lazysoundsystem@sd-44415 ~]$ docker start -ai test
/ # find /one -type f
/one/two/three/three
/one/two/three/four/four
/one/two/two
/one/one
/ #
```

```
[lazysoundsystem@sd-44415 ~]$ docker rm test
test
[lazysoundsystem@sd-44415 ~]$ docker create --name test -it -v /one/two alpine sh
c615606929c72b04118da8ecfbc9696c3308ca8645b5b36564e87e52c58ad00d
[lazysoundsystem@sd-44415 ~]$ docker cp one test:/
[lazysoundsystem@sd-44415 ~]$ docker start -ai test
/ # find /one -type f
/one/two/three/three
/one/two/three/four/four
/one/two/two
/one/one
/ #
```

**Comment by lazysoundsystem** [comment]
Apparently this is fixed in 1.12.1-rc1, so I'm closing this issue.
